extensions/src/main/java/org/apache/falcon/extensions/ExtensionBuilder.java

```diff
@@ -19,14 +19,52 @@
 package org.apache.falcon.extensions;
 
 import org.apache.falcon.FalconException;
+import org.apache.falcon.Pair;
 import org.apache.falcon.entity.v0.Entity;
+import org.apache.falcon.entity.v0.feed.Schema;
 
+import java.io.InputStream;
 import java.util.Properties;
 import java.util.List;
 
 /**
  * Extension interface to be implemented by all trusted and custom extensions.
  */
 public interface ExtensionBuilder {
-    List<Entity> getEntities(final String extensionName, final Properties extensionProperties) throws FalconException;
+
+    /**
+     * @param extensionName extension name.
+     * @param extensionStream stream comprising of the extension properties.
+     * @return List of the entities that are involved in the extension.
+     * @throws FalconException
+     */
+    List<Entity> getEntities(final String extensionName, final InputStream extensionStream) throws FalconException;
+
+    /**
+     * @param extensionName extension name.
+     * @return A graph comprising of the whole dag details of the extension.
+     * @throws FalconException
+     */
+    String getExtensionDescription(final String extensionName) throws FalconException;
+
+    /**
+     * @param extensionName extension name.
+     * @param extensionProperties Properties supplied will be validated.
+     * @throws FalconException
+     */
+    void validateExtension(final String extensionName, final Properties extensionProperties) throws FalconException;
+
+    /**
+     * @param extensionName extension name.
+     * @return List of the feed names along with the schema that the extension has generated if any.
+     * @throws FalconException
+     */
+    List<Pair<String, Schema>> getOutputSchemas(final String extensionName) throws FalconException;
+
+    /**
+     * @param extensionName extension name.
+     * @return Default template for the given extension.
+     * @throws FalconException
+     */
+    String getDefaultTemplate(final String extensionName) throws FalconException;
```

Comment:
Hmm.. This again, we thought we can just do a 'cat' of the template from the CLI. As long the extension packages it (which will be validated during register), CLI can access it.